tox.ini

```diff
@@ -1,18 +1,61 @@
+# This is a template `tox.ini` file for snaps
+# This file is managed by bootstack-charms-spec and should not be modified
+# within individual snap repos. https://launchpad.net/bootstack-charms-spec
+
 [tox]
-envlist = snap
-skipsdist = True
+skipsdist=True
+envlist = lint, unit, func
+skip_missing_interpreters = True
 
 [testenv]
-basepython = python3.5
-install_command = pip install {opts} {packages}
-passenv = HOME TERM
-whitelist_externals =
-    sudo
-    snapcraft
-
-[testenv:snap]
-deps = -r{toxinidir}/requirements.txt
+basepython = python3
+setenv = PYTHONPATH={toxinidir}
+
+[testenv:dev-environment]
+envdir = {toxinidir}/.venv
+deps =
+    pre-commit
+    {[testenv:lint]deps}
+    {[testenv:func]deps}
+commands =
+    pre-commit install
+
+[testenv:pre-commit]
+envdir = {[testenv:dev-environment]envdir}
+deps = {[testenv:dev-environment]deps}  # ensure that dev-environment is installed
+commands = pre-commit run --all-files
+
+[testenv:lint]
+commands =
+    pflake8
+    pylint --recursive=y .
+    black --check --diff --color .
+    isort --check --diff --color .
+deps =
+    black
+    flake8
+    pyproject-flake8
+    flake8-docstrings
+    pep8-naming
+    flake8-colors
+    colorama
+    isort
+    pylint
+    {[testenv:func]deps}
+
+[testenv:reformat]
+envdir = {toxworkdir}/lint
+deps = {[testenv:lint]deps}
+commands =
+    black .
+    isort .
+
+[testenv:unit]
```

Comment:
If we are going to have some code files in the snap, e.g. for extended cli options, it would be better to create an empty unittests directory and a sample command similar to func tests. Otherwise, if no unittest-worthy code are going to be created in the snap, I think we should remove unittests target from Makefile and here.

Reply:
the lint and unit workflows are tied together, thats the reason I had to include an empty tox env. i would also love to remove the unnecessary targets. if linting for functests are not necessary, I can get rid of it altogether and just apply test-functest and release workflows?

Reply:
Ah right, I forgot that the lint and unit tests are under the same workflow job. Personally, I suspect the snap will evolve to have at least one code file in the future, so I lean towards creating an empty unittest directory for now with a sample command and keep the target.

Reply:
I was thinking more like `{toxinidir}/pytest tests/unit ...`. But my opinion is not strong and its definitely not a blocker. I'm approving the PR.

Reply:
thanks for the approval and review. for the record, I tried that but an empty directory actually causes the pytest process to exit with non-zero return code so we need to add a dummy test, which I thought was a bit too much :sweat_smile: